web/js/codeworld.js

```diff
@@ -218,7 +220,68 @@ function initCodeworld() {
                 }
             },
             async: true
-        }
+        },
+        buttons: [
+            {
+                class: 'cw-button blue cw-button-save',
+                label: '',
+                title: 'Save',
+                callback: function (cm) {
+                    saveProject();
+                }
+            },
+            {
+                class: 'cw-button blue cw-button-search',
+                label: '',
+                title: 'Search',
+                callback: function (cm) {
+                    cm.execCommand('find')
+                }
+            },
+            {
+                class: 'cw-button blue cw-button-undo',
+                label: '',
+                title: 'Undo',
+                callback: function (cm) {
+                    cm.undo();
+                }
+            },
+            {
+                class: 'cw-button blue cw-button-redo',
+                label: '',
+                title: 'Redo',
+                callback: function (cm) {
+                    cm.redo();
+                }
+            },
+            {
+                class: 'cw-button blue cw-button-copy',
+                label: '',
+                title: 'Copy',
+                callback: function (cm) {
+                    window.clipboard = cm.getSelection();
+                    document.execCommand('copy');
+                }
+            },
+            {
+                class: 'cw-button blue cw-button-paste',
+                label: '',
+                title: 'Paste',
+                callback: function (cm) {
+                    cm.replaceSelection(window.clipboard);
```

Comment:
Does this mean that in addition to the system clipboard, you are keeping your own clipboard inside the web page?  That's going to cause lots of problems, such as making it hard to copy from another tab, an email, etc.  Any chance there's a way to request permissions we need and do this the right way?  Otherwise, maybe it's better to leave these off...

Reply:
useful resources which contains drafts how to do this https://www.lucidchart.com/techblog/2014/12/02/definitive-guide-copying-pasting-javascript/

https://jsfiddle.net/vtjnr6ea/ - looks like working example

I'm in progress with this.

Reply:
I can copy to clipboard, but can not copy from. According to this https://caniuse.com/#feat=clipboard we can not use copying from clipboard  via paste command in firefox. I can listen for `paste` event but not sure I can throw custom paste event on button click.

Reply:
Correction firefox should support paste with a focused editable field. In progress.

Reply:
I have no success with copying from system clipboard. Should I search for solution, or just delete "Past" button?